workers/data_refinery_workers/downloaders/array_express.py

```diff
@@ -123,14 +123,25 @@ def download_array_express(job_id: int) -> None:
         try:
             original_file = OriginalFile.objects.get(
                 source_filename=og_file['filename'], source_url=original_file.source_url)
-            original_file.is_downloaded = True
-            original_file.is_archive = False
-            original_file.absolute_file_path = og_file['absolute_path']
-            original_file.filename = og_file['absolute_path'].split('/')[-1]
-            original_file.calculate_size()
-            original_file.save()
-            original_file.calculate_sha1()
-            og_files.append(original_file)
+
+            # Sometimes a file needs to be redownloaded and processed,
+            # but if a file is part of an archive and then we requeue
+            # all of the files to be processed, we might end up
+            # reprocessing files. Therefore make sure that they
+            # actually needed to be downloaded before marking them as
+            # downloaded and queuing processor jobs.
+            if original_file.needs_downloading():
+                original_file.is_downloaded = True
+                original_file.is_archive = False
+                original_file.absolute_file_path = og_file['absolute_path']
+                original_file.filename = og_file['absolute_path'].split('/')[-1]
+                original_file.calculate_size()
+                original_file.save()
+                original_file.calculate_sha1()
```

Comment:
Saving before sha1 ? I think this will error, is this tested?

Reply:
Good catch. It didn't error out, but it definitely was getting covered by the end-to-end test for the array express downloader (I know cause I stuck a logging statement in there to make sure).